### ISSUE TITLE
ghc-9.4.1 illegal hardware instruction  stack runhaskell --package ansi-terminal

### ISSUE THREAD

**Issue by shayne-fletcher** [commercialhaskell/stack#5823]
### General summary

calling `setSGR` from `System.Console.Ansi` using ghc-9.4.1 from a program invoked by `stack runhaskell` abnormally terminates with error "illegal hardware instruction"

### Steps to reproduce

a reproducing script `test.sh` (say)
```bash
#!/usr/bin/env bash

cat > set-sgr-illegal.cabal <<EOF
cabal-version:      2.4
name:               set-sgr-illegal
version:            0.1.0.0
executable set-sgr-illegal
    main-is:          Main.hs
    build-depends:    base ^>= 4.17.0.0,ansi-terminal ^>= 0.11.3
    hs-source-dirs:   .
    default-language: Haskell2010
EOF

cat > stack.yaml <<EOF
resolver: ghc-9.4.1
extra-deps:
  - ansi-terminal-0.11.3
  - colour-2.3.6
EOF

cat > Main.hs <<EOF
import System.Console.ANSI
main = do
  setSGR [SetColor Foreground Dull Red]
EOF

stack runhaskell --package ansi-terminal Main.hs
```

### Expected

`stack` should build and execute the source file `Main.hs`; the exit code of `stack runhaskell` should be `0`.

### Actual

`stack runaskell` terminates abnormally with error message "illegal instruction" and exit code `132`.
 
### Stack version

```
$ stack --version
Version 2.7.5, Git revision ba147e6f59b2da75b1beb98b1888cce97f7032b1 x86_64 hpack-0.34.4
```
### Method of installation

* Official binary, downloaded from stackage.org or fpcomplete's package repository
* macOS monterey

**Comment by shayne-fletcher** [comment]
@mpilgrem g'day mike!

**Comment by shayne-fletcher** [comment]
it's *possible* https://www.mail-archive.com/ghc-devs@haskell.org/msg20589.html relates but, i did patch for it and it seems to work ok using just `cabal` alone.

**Comment by mpilgrem** [comment]
Can I ask, do you experience the same problem with the current version of Stack, built and installed from the `master` branch of this repository with (on macOS):
~~~
$ stack --stack-yaml=stack-macos.yaml install
~~~

**Comment by shayne-fletcher** [comment]
> Can I ask, do you experience the same problem with the current version of Stack, built and installed from the `master` branch of this repository with (on macOS):
> 
> ```
> $ stack --stack-yaml=stack-macos.yaml install
> ```

it appears so 😞 sorry!
```
❯ stack --version | head -n 1 && ./test.sh
Version 2.7.6, Git revision 966c74f8a54e9313adc8f9b891aa6d5035ee93c7 (8675 commits) RELEASE-CANDIDATE x86_64
est.sh: line 27: 84596 Illegal instruction: 4  stack runhaskell --package ansi-terminal Main.hs
132
```

**Comment by mpilgrem** [comment]
I've just pushed another GHC 9.4.1-related commit to the `master` branch, but I think this is unrelated to what that is intended to fix.

I think it may be an upstream with GHC 9.4.1, for the following reasons. (I am on Windows 11 and `mintty` is a dependency of `ansi-terminal` on Windows.)

First, I have no trouble in building an executable (on Windows 11) with:

~~~yaml
resolver: ghc-9.4.1
packages:
- .
extra-deps:
- ansi-terminal-0.11.3
- colour-2.3.6
- mintty-0.1.4

flags:
  mintty:
    win32-2-13-1: false
~~~

~~~haskell
module Main (main) where

import System.Console.ANSI

main :: IO ()
main = do
  setSGR [SetColor Foreground Dull Red]
  putStrLn "Red - a world about to dawn!"
~~~

and that works also with `resolver: ghc-9.2.4`.

With GHC 9.2.4, a 'raw' call to:

~~~
> stack --compiler ghc-9.2.4 exec -- runghc.exe -hide-all-packages -package"ansi-terminal-0.11.3" -package"colour-2.3.6" -packagebase -package"mintty-0.1.4" app/Main.hs
~~~
also works.

However. with GHC 9.4.1, a corresponding call to:

~~~
> stack --compiler ghc-9.2.4 exec -- runghc.exe -hide-all-packages -package"ansi-terminal-0.11.3" -package"colour-2.3.6" -packagebase -package"mintty-0.1.4" app/Main.hs
~~~

yields:

~~~
Access violation in generated code when executing data at 0x0

 Attempting to reconstruct a stack trace...

   Frame        Code address
Null address
~~~

**Comment by shayne-fletcher** [comment]
> I've just pushed another GHC 9.4.1-related commit to the `master` branch, but I think this is unrelated to what that is intended to fix.
> 
> I think it may be an upstream with GHC 9.4.1, for the following reasons. (I am on Windows 11 and `mintty` is a dependency of `ansi-terminal` on Windows.)
> 
> First, I have no trouble in building an executable (on Windows 11) with:
> 
> ```yaml
> resolver: ghc-9.4.1
> packages:
> - .
> extra-deps:
> - ansi-terminal-0.11.3
> - colour-2.3.6
> - mintty-0.1.4
> 
> flags:
>   mintty:
>     win32-2-13-1: false
> ```
> 
> ```haskell
> module Main (main) where
> 
> import System.Console.ANSI
> 
> main :: IO ()
> main = do
>   setSGR [SetColor Foreground Dull Red]
>   putStrLn "Red - a world about to dawn!"
> ```
> 
> and that works also with `resolver: ghc-9.2.4`.
> 
> With GHC 9.2.4, a 'raw' call to:
> 
> ```
> > stack --compiler ghc-9.2.4 exec -- runghc.exe -hide-all-packages -package"ansi-terminal-0.11.3" -package"colour-2.3.6" -packagebase -package"mintty-0.1.4" app/Main.hs
> ```
> 
> also works.
> 
> However. with GHC 9.4.1, a corresponding call to:
> 
> ```
> > stack --compiler ghc-9.4.1 exec -- runghc.exe -hide-all-packages -package"ansi-terminal-0.11.3" -package"colour-2.3.6" -packagebase -package"mintty-0.1.4" app/Main.hs
> ```
> 
> yields:
> 
> ```
> Access violation in generated code when executing data at 0x0
> 
>  Attempting to reconstruct a stack trace...
> 
>    Frame        Code address
> Null address
> ```

you are right @mpilgrem . i can reproduce the failure with just,
```
❯ runghc Main.hs
1]    15080 illegal hardware instruction  runghc Main.hs
```

i'll take the issue upstream!

**Comment by shayne-fletcher** [comment]
see https://gitlab.haskell.org/ghc/ghc/-/issues/22042

**Comment by mpilgrem** [comment]
GHC issue [#22042](https://gitlab.haskell.org/ghc/ghc/-/issues/22042) has been closed with the fix backported to GHC 9.4.2.
